### DISCUSSION TITLE
Implementing standardized off-diagonal Jacobian scalar assembly methods for IntegratedBC

### DISCUSSION THREAD

**Discussion by ttruster** [idaholab/moose#22091] [Q&A General] [answered]
As mentioned in my other thread here https://github.com/idaholab/moose/discussions/22020, I am working on implemented formulations for interface methods that require couplings to scalar variables. One currently published method is for weakly imposing periodic boundary conditions through surface integrals on opposing sides of the unit cell / RVE / microstructure; you can use either penalty, Nitsche, or Lagrange multiplier (mortar) approaches for this. This can provide benefit versus the `HomogenizationConstraintScalarKernel` MOOSE system because the DOF coupling is only on the external surfaces rather than every quadrature point through the material model. In the case of PCG iterative solver, solutions can be obtained WITHOUT preconditioning in a <1000 iterations for a 45,000 DOF system Jacobian.

After scouring the framework source, I have found that most objects besides the `Kernel` do not have implementations of the off-diagonal coupling with scalar variables (which is the holder for the macroscale, volume-averaged quantity). More importantly, the assembly tagging system for scalar variables is missing. For existing objects like `ScalarLagrangeMultiplier`, the assembly is handled in a non-standard way.

So it seems like now is a great time to develop some standards for computing and assembling the scalar contributions, while allowing a developer to still derive off of higher base classes and avoid these standards if their problem doesn't fit the usual mold.

Since each of the IntegratedBC, (Mortar)Constraint, InterfaceKernel, and DGKernel are in need of such Jacobian and assembly implementations, I'm volunteering to tackle them in that order, since I can formulate diffusion equation analogs for the first 3 of the 4.

To finalize the objectives of the developments, I don't know if that discussion should happen within an issue or through this forum discussion. So I'll just post 4 thoughts to start a dialog to finalize on the interface used for treating these scalar variables with other variables:
1. I've also noticed objects besides `Kernel` and `DGKernel` are missing pre-calculate method calls to allow the override of the `ResidualObject` method prior to the quadrature loop. Technically most classes can override the lower level quadrature and add it back, but the mortar constraint does a 'final' override which prevents this (why?). So I'll add those along the way.
2. I need help understanding how the assembly sequence works (prepare, tagging, adding, caching, accumulating)
3. I propose that the off-diagonal Jacobian scalar methods always provide two quadrature loops which call methods with a name like "computeQpOffDiagJacobianVariableScalar" and "computeQpOffDiagJacobianScalarVariable" to handle both of the directions of the coupling (regular variable to scalar variable and scalar variable to regular variable). Generally, these two couplings are not equal (equal for `ScalarLagrangeMultiplier`, not for `HomogenizedTotalLagrangianStressDivergence`). Handling these contributions in each regular variable object class (Kernel, …) avoids having to make a bunch of different loops within sub-classes of `ScalarKernel`.
4. Let user objects handle the integral of the residual and on-diagonal Jacobian contributions of the scalar variable, as in `HomogenizationConstraintIntegral`. Providing template user objects for each regular variable object class along with one scalar kernel that references the residual/Jacobian would seem like a "friendly" system for new developers to contribute to

Thoughts???

**Comment by GiudGiud** [comment, marked as answer]
@lindsayad could you please answer Pr Truster's questions
I think I can learn from this too

**Reply by GiudGiud** [reply]
ok seems everyone is chatting there already
https://github.com/idaholab/moose/issues/22174
https://github.com/idaholab/moose/discussions/22020

**Reply by lindsayad** [reply]
yea let's consolidate discussion in #22174 and the PR addressing it, #22246

**Reply by GiudGiud** [reply]
ok closing this then.
